File: ArraysI/NextPermutation.py
class Solution:
    def nextPermutation(self, nums: list[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        
        n  = len(nums) - 1
        breakPoint = -1

        for i in range(n - 1, -1, -1):
            if nums[i] < nums[i + 1]:
                breakPoint = i
                break

        if breakPoint == -1:
            nums.reverse()
            return

        for i in range(n, breakPoint, -1) :
            if(nums[i] > nums[breakPoint]) :
                nums[i], nums[breakPoint] = nums[breakPoint], nums[i]
                break


        i, j = breakPoint + 1, n 

        while i <= j :
            nums[i], nums[j] = nums[j], nums[i]
            i+=1
            j-=1


        return nums

File: Arrays/TwoPointers/NextPermutation.cpp
#include <bits/stdc++.h>
using namespace std;

namespace next_permutation {

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        
        int n = nums.size() - 1;
        int breakPoint = -1;

        for (int i = n - 1; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                breakPoint = i;
                break;
            }
        }

        if (breakPoint == -1) {
            reverse(nums.begin(), nums.end());
            return;
        }

        for (int i = n; i > breakPoint; i--) {
            if (nums[i] > nums[breakPoint]) {
                swap(nums[i], nums[breakPoint]);
                break;
            }
        }

        int i = breakPoint + 1;
        int j = n;

        while (i <= j) {
            swap(nums[i], nums[j]);
            i++;
            j--;
        }
    }
};

} // namespace next_permutation